### PR TITLE
8293841: RISC-V: Implementation of Foreign Function & Memory API (Preview)

### PR DESCRIPTION
**Pull request by feilongjiang**
Add experimental Foreign Function & Memory API support for RISC-V.

For details of the FFM API RISC-V port please refer to [JBS issue](https://bugs.openjdk.org/browse/JDK-8293841)

Testing:

- [x] jdk_foreign with release/fastdebug build on linux-riscv64
- [x] jdk_foreign with release/fastdebug build on linux-x64
- [x] jdk_foreign with release/fastdebug build on linux-aarch64

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8293841](https://bugs.openjdk.org/browse/JDK-8293841): RISC-V: Implementation of Foreign Function & Memory API (Preview)

### Reviewers
 * [Jorn Vernee](https://openjdk.org/census#jvernee) (@JornVernee - **Reviewer**) ⚠️ Review applies to [fee96ae8](https://git.openjdk.org/jdk/pull/11004/files/fee96ae817edc7b8cfdd5c431edf0b44ac6775c5)
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**) ⚠️ Review applies to [c4809317](https://git.openjdk.org/jdk/pull/11004/files/c4809317a83a246b2bc7f41cea6e2039b997d1e7)
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**) ⚠️ Review applies to [c4809317](https://git.openjdk.org/jdk/pull/11004/files/c4809317a83a246b2bc7f41cea6e2039b997d1e7)
 * [Yadong Wang](https://openjdk.org/census#yadongwang) (@yadongw - Author)

### Contributors
 * Weikai He `<weikai@isrc.iscas.ac.cn>`
 * Fei Yang `<fyang@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11004/head:pull/11004` \
`$ git checkout pull/11004`

Update a local copy of the PR: \
`$ git checkout pull/11004` \
`$ git pull https://git.openjdk.org/jdk pull/11004/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11004`

View PR using the GUI difftool: \
`$ git pr show -t 11004`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11004.diff">https://git.openjdk.org/jdk/pull/11004.diff</a>

</details>

### CODE REVIEW

**Review by JornVernee — COMMENTED** [review]
Excellent. No Comments.

I am running some sanity testing on our CI, and will approve when it comes back green.

**Review by JornVernee — APPROVED** [review]

File: test/jdk/java/foreign/callarranger/TestRISCV64CallArranger.java [hunk 1]
```diff
@@ -0,0 +1,542 @@
+/*
+ * Copyright (c) 2020, 2022, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2023, Institute of Software, Chinese Academy of Sciences. All rights reserved.
```
Comment (RealFYang) [inline test/jdk/java/foreign/callarranger/TestRISCV64CallArranger.java:3]: I think we should also break this into two separate lines like you do for other newly-added files.

**Review by RealFYang — CHANGES_REQUESTED** [review]

File: test/jdk/java/foreign/callarranger/TestRISCV64CallArranger.java (hunk 1 above)
Comment (feilongjiang) [reply test/jdk/java/foreign/callarranger/TestRISCV64CallArranger.java]: Oops! Missed here, fixed.

**Review by RealFYang — APPROVED** [review]
Looks good to me. This also passed tier1-4 tests on my HiFive Unmatched board.

**Comment by feilongjiang** [comment]
> Excellent. No Comments.
> 
> I am running some sanity testing on our CI, and will approve when it comes back green.

@JornVernee, thanks for the reviewing and testing!

**Review by shipilev — APPROVED** [review]
Looks okay from the cursory review.

File: src/hotspot/cpu/riscv/codeBuffer_riscv.cpp [hunk 2]
```diff
@@ -1,6 +1,7 @@
 /*
  * Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
- * Copyright (c) 2022, Institute of Software, Chinese Academy of Sciences. All rights reserved.
+ * Copyright (c) 2022, Institute of Software, Chinese Academy of Sciences.
```
[the remark this answers is not in the document]
Comment (feilongjiang) [reply src/hotspot/cpu/riscv/codeBuffer_riscv.cpp:3]: The copyright of ISCAS is too long for one line, so we split them into separate lines.

**Review by yadongw — APPROVED** [review]
lgtm

**Comment by feilongjiang** [comment]
@JornVernee @RealFYang @shipilev @yadongw -- Thank you all for the review! I'm going to integrate it then.
/integrate

**Pull request closed without merging**